// @beisen/approve/webpack.dev.config.js
var path = require("path");
var webpack = require("webpack");

module.exports = {
	entry : './index.js',
	module : {
		loaders : [
			{
				loader : 'babel-loader',
				test : /\.js$/,
				exclude : /node_modules/
			},
			{
				test: /\.scss$/,
				loader: "style!css!sass"
			},
			{
				test: /\.(jpe?g|png|gif)$/i,
				loader: 'file-loader'
			},
			{ 
              	test: /\.woff(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=30000&name=[name]-[hash].[ext]" 
          	},
          	{   
              	test: /\.woff2(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=30000&name=[name]-[hash].[ext]" 
          	},
          	{   
              	test: /\.ttf(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=30000&name=[name]-[hash].[ext]" 
          	},
          	{ 
              	test: /\.eot(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=30000&name=[name]-[hash].[ext]" 
          	},
          	{ 
              	test: /\.svg(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=10000&mimetype=image/svg+xml" 
          	}
		]
	},
	output: {
		filename: 'main.bundle.js'
	},
	devServer : {
		filename : 'main.bundle.js',
		host : '0.0.0.0',
		port : 8080
	},
	resolve: {
	    modulesDirectories: [
	      'src',
	      'node_modules'
	    ],
	    extensions: ['', '.json', '.js', '.jsx']
  	}
}
