// @beisen/approve/webpack.config.js
var path = require("path");
module.exports = {
	entry : './src/index.js',
	module : {
		loaders : [
			{
				loader : 'babel-loader',
				test : /\.js$/,
				exclude : /node_modules/
			},
			{
				test: /\.scss$/,
				loader: "style!css!sass"
			},
			{
				test: /\.(jpe?g|png|gif)$/i,
				loader: 'file-loader'
			},
			{ 
              	test: /\.woff(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=30000&name=[name]-[hash].[ext]" 
          	},
          	{   
              	test: /\.woff2(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=30000&name=[name]-[hash].[ext]" 
          	},
          	{   
              	test: /\.ttf(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=30000&name=[name]-[hash].[ext]" 
          	},
          	{ 
              	test: /\.eot(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=30000&name=[name]-[hash].[ext]" 
          	},
          	{ 
              	test: /\.svg(\?v=\d+\.\d+\.\d+)?$/, 
              	loader: "url?limit=10000&mimetype=image/svg+xml" 
          	}
		]
	},
	output: {
	    library: 'index',
	    libraryTarget: 'umd'
	},
	resolve: {
	    modulesDirectories: [
	      'src',
	      'node_modules'
	    ],
	    extensions: ['', '.json', '.js', '.jsx']
		},
	externals: {
		'@beisen/ethos': {
			root: '@beisen/ethos',
			commonjs2: '@beisen/ethos',
			commonjs: '@beisen/ethos',
			amd: '@beisen/ethos'
		},
	}
}
